File: fastify/src/lib/redis.ts
import Redis from "ioredis";
import { env } from "../config/env";
import { logger } from "./logger";

export let redisClient: Redis | null = null;

if (env.REDIS_URL) {
  redisClient = new Redis(env.REDIS_URL, { lazyConnect: true, enableReadyCheck: false });
  redisClient.on("error", (err) => logger.warn("Redis error", { error: err.message }));
}

const TTL = 300;

async function get<T>(key: string): Promise<T | null> {
  if (!redisClient) return null;
  try {
    const val = await redisClient.get(key);
    return val ? (JSON.parse(val) as T) : null;
  } catch {
    return null;
  }
}

async function set(key: string, value: unknown, ttl = TTL): Promise<void> {
  if (!redisClient) return;
  try {
    await redisClient.set(key, JSON.stringify(value), "EX", ttl);
  } catch {
    // non-fatal
  }
}

async function del(key: string): Promise<void> {
  if (!redisClient) return;
  try {
    await redisClient.del(key);
  } catch {
    // ignore
  }
}

export const cache = {
  async ping(): Promise<boolean> {
    if (!redisClient) return false;
    try {
      return (await redisClient.ping()) === "PONG";
    } catch {
      return false;
    }
  },

  getLinkByCode: (shortCode: string) => get(`link:code:${shortCode}`),
  setLinkByCode: (shortCode: string, val: unknown) => set(`link:code:${shortCode}`, val),
  invalidateLinkByCode: (shortCode: string) => del(`link:code:${shortCode}`),

  getUserLinks: (userId: string) => get(`user:links:${userId}`),
  setUserLinks: (userId: string, val: unknown) => set(`user:links:${userId}`, val),
  invalidateUserLinks: (userId: string) => del(`user:links:${userId}`),

  getPublicUser: (username: string) => get(`public:user:${username}`),
  setPublicUser: (username: string, val: unknown) => set(`public:user:${username}`, val),
  invalidatePublicUser: (username: string) => del(`public:user:${username}`),

  getPublicLinks: (username: string) => get(`public:links:${username}`),
  setPublicLinks: (username: string, val: unknown) => set(`public:links:${username}`, val),
  invalidatePublicLinks: (username: string) => del(`public:links:${username}`),

  async invalidateAll({ linkCode, userId }: { linkCode?: string; userId?: string }): Promise<void> {
    const tasks: Promise<void>[] = [];
    if (linkCode) tasks.push(del(`link:code:${linkCode}`));
    if (userId) tasks.push(del(`user:links:${userId}`));
    await Promise.all(tasks);
  },
};

File: fastapi/app/lib/redis.py
import json

import redis as _redis

from ..config import settings

_client: _redis.Redis | None = None
_initialized = False

_TTL = {'LINK_BY_CODE': 60, 'PUBLIC_USER': 120, 'PUBLIC_LINKS': 60, 'USER_LINKS': 30}


def _get_client() -> _redis.Redis | None:
    global _client, _initialized
    if _initialized:
        return _client
    _initialized = True
    if not settings.REDIS_URL:
        return None
    try:
        _client = _redis.from_url(settings.REDIS_URL, decode_responses=True,
                                   socket_connect_timeout=2, socket_timeout=2)
    except Exception:
        pass
    return _client


def _get(key: str):
    r = _get_client()
    if not r:
        return None
    try:
        val = r.get(key)
        return json.loads(val) if val else None
    except Exception:
        return None


def _set(key: str, value, ttl: int) -> None:
    r = _get_client()
    if not r:
        return
    try:
        r.setex(key, ttl, json.dumps(value))
    except Exception:
        pass


def _del(*keys: str) -> None:
    r = _get_client()
    if not r or not keys:
        return
    try:
        r.delete(*keys)
    except Exception:
        pass


def get_link_by_code(code): return _get(f'link:code:{code}')
def set_link_by_code(code, val): _set(f'link:code:{code}', val, _TTL['LINK_BY_CODE'])
def invalidate_link_by_code(code): _del(f'link:code:{code}')

def get_public_user(username): return _get(f'public:user:{username}')
def set_public_user(username, val): _set(f'public:user:{username}', val, _TTL['PUBLIC_USER'])

def get_public_links(username): return _get(f'public:links:{username}')
def set_public_links(username, val): _set(f'public:links:{username}', val, _TTL['PUBLIC_LINKS'])

def get_user_links(user_id): return _get(f'user:links:{user_id}')
def set_user_links(user_id, val): _set(f'user:links:{user_id}', val, _TTL['USER_LINKS'])
def invalidate_user_links(user_id): _del(f'user:links:{user_id}')

def invalidate_all(*, link_code=None, user_id=None):
    keys = []
    if link_code:
        keys.append(f'link:code:{link_code}')
    if user_id:
        keys.append(f'user:links:{user_id}')
    if keys:
        _del(*keys)

def ping() -> bool:
    r = _get_client()
    if not r:
        return False
    try:
        return bool(r.ping())
    except Exception:
        return False
